content/5_Plat/Centroid.mdx

```diff
@@ -235,99 +235,142 @@ private static class Centroid {
 
 ## Solution - Xenia & Tree
 
+<!-- code from https://codeforces.com/contest/342/submission/114241456 (with a couple of minor modifications) -->
+
 ```cpp
-#include <cstdio>
-#include <cstring>
+#include <iostream>
+#include <algorithm>
 #include <vector>
-
-template<typename T> bool ckmin(T& a, const T& b){return b<a?a=b,1:0;}
-
-const int MN = 1e5+10, INF = 0x3f3f3f3f;
-
-int N, M, s[MN], m[MN][2], t, b, d;
-bool r[MN], red[MN];
-std::vector<int> a[MN], v[MN];
-struct info
-{
-	public:
-		int n,b,d;
-};
-std::vector<info> g[MN];
-
-int dfs(int n, int p=0)
-{
-	s[n]=1;
-	for(auto x:a[n])
-		if(x!=p&&!r[x])
-			s[n]+=dfs(x, n);
-	return s[n];
+#pragma GCC optimize ("Ofast")
+#pragma GCC target ("avx2")
+#pragma GCC optimize("unroll-loops")
+using namespace std;
+ 
+const int inf = 2000000011;
+#define PB push_back
+ 
+int N, Q, W[100005], P[100005], ans[100005];
+vector<int> adj[100005];
+bool col[100005];
+
+// W[] is weight(size) of subtree (excluding deleted nodes)
+// P[] is parent in centroid tree
+// ans[] is closest red node in subtree of centroid tree
+// adj[] stores original tree given in input
+// col[] stores if node has already been colored red (optional)
+ 
+bool r[100005];
+// r[] stores if a node has been deleted
+ 
+int dfs_w(int u, int p = 0) {
+    if (r[u])
+        return 0;
+    W[u] = 1;
+    for (int v : adj[u])
+        if (v != p && !r[v])
+            W[u] += dfs_w(v, u);
+    return W[u];
 }
-int find(int n, int ms, int p=0)
-{
-	for(auto x:a[n])
-		if(!r[x]&&x!=p&&s[x]*2>ms)
-			return find(x, ms, n);
-	return n;
+ 
+int centroid(int u, int n, int p = 0) {
+    for (int v : adj[u])
+        if (v != p && !r[v])
+            if (W[v] * 2 > n) 
+                return centroid(v, n, u);
+    return u;
 }
-void dfs2(int n, int p=0)
-{
-	for(auto x:a[n])
-		if(!r[x]&&x!=p)
-			++d, dfs2(x, n), --d;
-	g[n].push_back({t,b,d});
+ 
+void build(int u = 1, int p = 0) {
+    int c = centroid(u, dfs_w(u));
+    P[c] = p;
+    r[c] = 1;
+    for (int v : adj[c])
+        if (!r[v]) 
+            build(v, c);
 }
-void centroid(int n=1)
-{
-	n = find(n, dfs(n));
-	v[n].reserve(a[n].size());
-	for(auto x:a[n])
-		if(!r[x])
-		{
-			t=n, b=v[n].size(), d=1;
-			dfs2(x, n);
-			v[n].push_back(INF);
-		}
-	r[n]=1;
-	for(auto x:a[n])
-		if(!r[x])
-			centroid(x);
+ 
+// this stuff is for tree distance, see binary jumping mondule
```

Comment:
there's no need to use binary jumping

Reply:
How would you do it without binary jumping?

Reply:
solution on the left doesn't use binary jumping

Reply:
Both tutorials linked on the article use the binary jumping/lca method, so I think its more intuitive

Reply:
oops started this comment a few days ago but forgot to submit it

Reply:
I'll work on the implementation, but I'm quite busy the next few days due to schoolwork.